zookeeper-server/src/main/java/org/apache/zookeeper/server/PrepRequestProcessor.java

```diff
@@ -1010,9 +1011,11 @@ private void validateCreateRequest(String path, CreateMode createMode, Request r
             if (id == null || id.getScheme() == null) {
                 throw new KeeperException.InvalidACLException(path);
             }
-            if (id.getScheme().equals("world") && id.getId().equals("anyone")) {
+            if (id.getScheme().equals("world") && id.getId().equals("anyone") && !QuorumPeerConfig
+                .isSetX509ClientIdAsAclEnabled()) {
                 rv.add(a);
-            } else if (id.getScheme().equals("auth")) {
+            } else if (id.getScheme().equals("auth") || QuorumPeerConfig
```

Comment:
So my understanding is that we are creating another option to set up how ACL will be managed. And it is overriding the existing options. My question is that is it necessary? If users specify the id schema as "auth" as they are supposed to do it, then do we need to add this additional config?

I guess it might be the case that X509 Id is not compatible with the "auth" id schema. Please let me know if this is the case?

Reply:
Yes, this could be done at creation time by clients. But we want to do this on the server-side, hence the feature (with a configurable option). This is because there are many different client implementations, and doing this on the server-side is much easier to control the ACL behavior.

I'm not sure what you're referring to X509 not being compatible with "auth".

Reply:
I got the idea. I was not thinking about the need of enforcing it on the server-side.

But the 2nd half of my question is not answered. With this change, if we enable setX509ClientIdAsAcl, then we actively disable "world" acl for "anyone". Moreover, a client won't be able to create a new node with someone else's Id. Because we will always use authInfo.
Is it necessary? IMO, the client should control if their data needs to be protected and how.

And even I put one step back, would it be easier and cleaner if we just add a config to shut off other options like "world"/"anyone" and throw clear exceptions.
The problem of this design is, for example, a ZK client in service A used to create a new node with service B's Id. Then when we turn this feature on our ZK server-side, the behavior changed. But service A won't be aware. The operation will be done successfully. It may take a long time before service B notices that it cannot read the node.

Reply:
@jiajunwang IMO the creator of the node owns the node is natural. If someone else should also have access to the node, the creator can set acl for the node after creation. We will inform our users that the pattern now is changed and they should take care of the acl if other services need to access the nodes. Moreover, we have categorized our users' access pattern into four categories, only one of them (open read access) will have the problem as you described, and for that category, we will set all the nodes that they create as (world, anyone, r). For more details, you can check out the wiki about cross-domain clients.

Reply:
My concern is not about use cases themselves. But the protocol becomes ambiguous with this change. I would suggest failing the unsupported request if you are going to override the owner info anyway. Don't let users guess and assume according to some documents (since there is no way to enforce a user to read a document before they use the API).

Reply:
Synced offline with Hunter and Molly. So the world/anyone is the default setup if the user did not put anything in the request. Meaning there is no way to distinguish between an intentional world/anyone or default no acl request. Given this, it makes the usage much harder if we are going to fail the request with no auth specified. And the current methodology is practical.

The only downside is that if any user intentionally specifies world/anyone, that intention would be overwritten silently.

Comparing these options, I think the current direction tradeoffs better.

Reply:
Explained the way Zk clients set ACLs to Jiajun. Resolving.